design/ShaderLanguageFAQ.md

```diff
@@ -0,0 +1,157 @@
+or “WGSL” for short.
+
+It is under development, but will have the following traits:
+
+*   It is a text-based language.
+*   It is suitable for humans to read and write.
+*   It is strongly typed.
+*   It directly represents block-structured programming, including
+    sequence, selection (if/else), and repetition (looping).
+*   It can be faithfully and simply converted to and from SPIR-V
+    “Shader” modules that only use GPU features available in WebGPU.
+
+The W3C WebGPU community group agreed to this direction during the
+February 2020 face-to-face meeting in Redmond.
+
+
+## How do you pronounce “WGSL”?
+
+The W3C WebGPU Community Group does not prescribe a particular
+pronunciation of “WGSL”.
+
+You can say it letter-by-letter, like “W G S L”.
+
+You can say it as a single word, like “wiggzle”, “wiggles”, or “wigsel”.
+
+
+## What does a WGSL shader look like?
+
+WGSL has not yet been specified.
+However, early prototyping work indicates the language can have a look
+and feel as represented in the following examples:
+
+A simple fragment shader:
+
+
+```
+[[location 0]] var<out> gl_FragColor : vec4<f32>;
+
+fn main() -> void {
+    gl_FragColor = vec4<f32>(0.4, 0.4, 0.8, 1.0);
+    return;
+}
+entry_point fragment = main;
+```
+
+
+A code fragment showing if/else:
+
+
+```
+  if (value > 5) {
+    wvar = 14;
+  } elseif (value > 2) {
+    wvar = 12;
+  } else {
+    wvar = 0;
+  }
+```
+
+
+A code fragment showing a switch:
+
+
+```
+switch (a) {
+  case 1: {
+    # The break is implicit
+  }
+  case 2: {
+     fallthrough;
+  }
+  default: {
+  }
+}
+```
+
+
+A code fragment showing a loop:
+
+
+```
+var i : i32 = 0;
+var b : f32 = 2.4;
+loop {
+  break if (i == 4);
+
+  b = b * 3.2;
+  continuing {
+     i = i + 1;
+  }
+}
+```
+
+
+
+## Where can I find the spec?
+
+The WGSL spec will be developed on the WebGPU Community Group area on GitHub.
+See https://github.com/gpuweb/gpuweb/tree/master/wgsl
+
+
+## What browsers support WGSL?
+
+WGSL is the shader language for WebGPU.
+Any browser implementing WebGPU will be required to accept shaders in WGSL.
+Specifically, the WebGPU conformance test suite will use shaders written
+in WGSL.
+
+Browser support for WGSL is expected to be developed over time, along
+with the development of WGSL itself.
+
+
+## Will browsers accept other shader languages for WebGPU?
+
+Accepting a shader language other than WGSL is not part of WebGPU.
+Doing so harms interoperability, as a shader in that other language is
+not portable across browsers.
+
+
+## Where can I find WGSL tooling?
+
+The WebGPU Community Group does not deliver tooling for WGSL.
+However, WebGPU implementers are expected to develop tooling
+for handling WGSL, including parsing, validation, and compilation.
+
+At this time, we expect at least three essentially independent
+implementations of WGSL ingestion as part of browser support for WebGPU.
+Independent implementation serves as an important check on the quality
```

Comment:
"serves as a check on the quality"? Doesn't that technically indicate that independent implementations reduce quality? (as in "checks and balances")

Reply:
I don't think so? It's saying that having multiple implementations means we know the spec can be implemented and isn't just a product of a single compiler. Those implementations check that the spec is actually correct, vs what a given implementation happens to do.

Reply:
That's to contrast with WebGL which is a compiler monoculture now.

Reply:
Right, I agree with the intention, but I believe the specific words in this document do not match the intention.

Reply:
WebGL is not a compiler monoculture, and it's actually a major problem for users. Safari's WebGL GLSL parser has bugs I've worked around on past projects. And Chrome-on-Android does not go through ANGLE's shader translator (best I know?), leading to me needing to work around plenty of driver issues in my WebGL applications.

The graphics community has years of experience with the "multiple reimplementations; no reference tooling" state of affairs with GLSL. One of Vulkan's successes in this arena was developing reference tooling, supported by Khronos themselves.

Reply:
I think the words say what I intend them to say. 

Independent implementation ensures that multiple people have looked at the spec very closely, i.e. close enough to implement it.  Doing so shines a light on lurking ambiguity or gaps in the spec.  It's not a exhaustive proof, but "an important check".  The other line of defense is the test suite.  Hopefully independent implementation will induce ambiguity-reducing discussion, and when that's resolved the decision is baked into both a spec update and a test update.

That's my intent here, but I did't want to write so many words for it.

Reply:
In typical English idiom, "serves as an important check on X" means placing a limitation on X or reducing X, rather than providing evidence for X. c.f. "Congress serves as an important check on the power of the presidency." Perhaps it can have that other meaning too, but in that case why use an ambiguous phrase? You could say "Independent implementation provides evidence for the quality of the spec."

This may seem nitpicky but use of ambiguous wording like this can be legitimately confusing to some people (e.g. non-native English speakers or the neurodiverse).